Implementation of RegexMappableResultSet

thomasnield opened the pull request:
Allow the mapping of getters on column names with regular expressions. Discussed in #34.

thomasnield commented:
Shoot, I don't know why the Travis build failed...

davidmoten commented on src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:
```diff
@@ -0,0 +1,830 @@
+                columnLabels[i-1] = rs.getMetaData().getColumnName(i);
+            }
+
+        } catch (Exception e) {
```
Can you narrow this to only the thrown types pls?

thomasnield replied:
Sure.

davidmoten commented on src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:
```diff
@@ -0,0 +1,830 @@
+            }
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
I think I've got a SQLRuntimeException to wrap e with here

davidmoten commented on src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:
```diff
@@ -0,0 +1,830 @@
+    private final String[] columnLabels;
+    private final HashMap<String,Integer> cachedIndices = new HashMap<>(); //-1 means unfound
+
+    private RegexMappableResultSet(ResultSet rs) {
```
Happy to have a private constructor but I can't see a static factory method to create one of these

davidmoten commented on src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:
```diff
@@ -0,0 +1,830 @@
+                            return Integer.valueOf(i);
+                        }
+                }
+            return Integer.valueOf(-1);
```
If you replace this with a `SQLException` throw then users may get meaningful error messages when col pattern doesn't match anything. As it stands they will get "index -1 is an illegal argument" or something like that. The `findColumn` method by the `ResultSet` specification should also return a SQLException if column not found (rather than returning -1).

thomasnield replied:
Okay will do.

thomasnield replied:
Do you think there's any value to optionally make column mappings optional and be able to return null rather than a SQLException? I sometimes find it helpful with disparate data sources where fields may or may not be there, and be able to make the return value null rather than an exception.

davidmoten replied:
Interesting, you could do that if you like. A boolean parameter in the factory or a separate ResultSet implementation that you compose with this one (that would be nicer methinks so you could have that even when you didn't want pattern matches on columns).

thomasnield replied:
Okay I'll implement that later today. If I do two implementations I may have to revise the exception strategies for unfound columns on `RegexMappableResultSet` otherwise it would logically clash with this `NullableIfUnfoundResultSet`. My reasoning so far indicates the `RegexMappableResultSet` must have some notion of nullability or else the getters() will only throw exceptions if columnLabels are unfound.

thomasnield replied:
Darn, how do I deal with primitives in this then... since they can't be null. Default primitive value policy?

thomasnield replied:
Never mind... I just looked through the `ResultSet` documentation and found the numeric primitive getters return 0 if null, and `getBoolean()` returns false if null. Therefore I'll make the `NullableIfUnfoundResultSet` do the same.

thomasnield commented:
Okay did my first successful test. RegexMappableResultSet is working. Of course I started the index at 0, not 1, to retrieve the column headers so I fixed that.

thomasnield commented:
I did a test using a SQLite database. I'll create a unit test later if I get a chance to see how your tests are structured. 

Do we really want the factory/constructor to throw a SQLException? It feels kind of awkward in the monad.

```
public static void main(String[] args) {
        try {
            Database.from(DriverManager.getConnection("jdbc:sqlite:C:\\Users\\Thomas\\OneDrive\\Data\\finance_assistant.db"))
                    .select("SELECT * FROM CATEGORIES")
                    .resultSetTransform(rs -> {try { return RegexMappableResultSet.of(rs);} catch (Exception e) { e.printStackTrace(); throw new RuntimeException(e);}})
                    .get(rs -> rs.getInt(".*ID.*"))
                    .subscribe(i -> System.out.println(i));

        }
        catch (Exception e){
            e.printStackTrace();
        }
    }
```

davidmoten commented:
I reckon you don't want the factory method to throw SQLException, it's just annoying. Get it to wrap the SQLException with an unchecked SQLRuntimeException.

By the way `Checked` class in rxjava-extras https://github.com/davidmoten/rxjava-extras  offers convenience methods like this if you want to do this elsewhere:

```java
.resultSetTransform(Checked.f1(rs -> RegexMappableResultSet.of(rs)));
```

davidmoten commented:
I think `RegexMappableResultSet` is a beaut name. However, I'd go for `RegexMappableResultSet.from(rs)` for the factory method. `of` doesn't read right and it's a method I'd use for say collection creation (make a set "of" a, b) and we are building a RegexMappableResultSet "from" a ResultSet.

davidmoten commented:
As far as tests go, we just need a standalone junit test class. You can just mock the input result set.

davidmoten commented on src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:
```diff
@@ -0,0 +1,834 @@
+package com.github.davidmoten.rx.jdbc;
+
+
+import org.sqlite.SQLiteConnection;
```
this will cause the build to fail

thomasnield replied:
oops... my bad.

davidmoten commented:
To add a real test with a database add something into `DatabaseTestBase` which has lots of examples to follow.

thomasnield commented:
Cool! I was wondering if there was a library somewhere like `Checked` to accomplish that. Good to know!

davidmoten commented:
Doesn't compile:

```
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[29,9] 'catch' without 'try'
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[20,9] 'try' without 'catch', 'finally' or resource declarations
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[34,5] illegal start of expression
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[34,12] illegal start of expression
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[34,41] ';' expected
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[34,56] ')' expected
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[34,59] illegal start of expression
[ERROR] /home/travis/build/davidmoten/rxjava-jdbc/src/main/java/com/github/davidmoten/rx/jdbc/RegexMappableResultSet.java:[34,60] ';' expected
```

thomasnield commented:
That's what happens when you edit the files on GitHub directly and don't compile them locally. One moment...

thomasnield commented:
Okay done, sorry about that.

thomasnield commented:
Not quite sure why Travis build failed... I can't find anything Java 8-specific...

thomasnield commented:
Just finished `NullableIfUnfoundResultSet`. I couldn't find a more kosher way to do this, so the wrapper will attempt to retrieve the `findColumn()` index. If an exception is thrown, it catches it and caches it as a -1, which indicates the `columnLabel` is not present. 

Fortunately it will only do this try-catch a few times and then it will rely on the cache after the first record is processed.

The primitive getters use the same "null" primitive defaults as a standard `ResultSet`, and will use them for unfound columns. `getInt()` and other numeric getters will return 0, and `getBoolean()` will return false. 

I have not tested this yet. Still need to compose the test cases.

thomasnield commented:
Fixed Java 6/7 compile issues. Was too busy looking for lambdas but then I figured out how to use the Travis log. Found some new Java 8-specific methods added to `ResultSet` interface. Removed them and now everything is good to go.

davidmoten commented:
Can I suggest that when a column is not found then `RegexMappableResultSet` should throw a `ColumnMatchNotFoundException` which would extend `SQLException` and you would create. Then your  `NullableIfUnfoundResultSet` might only return null if the exception it detected was an instance of `ColumnMatchNotFoundException` otherwise it would rethrow the exception.

davidmoten commented:
By the way `NullableIfUnfoundResultSet` should probably be called something like `NullIfNotMatchedResultSet` because Nullable implies something can be set  to null and this is about reading values.

thomasnield commented:
All the above makes perfect sense. I'll make those revisions tomorrow morning Dallas time. And I'll look at getting the unit tests in.

thomasnield commented:
Are you considering limiting the API to `resultSetTransform()`? If so that's okay because I can just use my `NullIfNotMatchedResultSet` and `RegexMappableResultSet` locally. The `resultSetTransform()` fulfills my needs beautifully.

thomasnield commented:
I open-sourced the `RegexMappableResultSet` as a separate library. Figured the `resultSetTransform()` works great with any custom `ResultSet` implementation, including that one. https://github.com/thomasnield/regex-jdbc
